pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1136 @@
+		if !apierrors.IsServerTimeout(err) {
+			klog.Errorf("CIDR assignment for node %q failed: %v. Releasing allocated CIDR", node.Name, err)
+			for _, cidr := range data.allocatedCIDRs {
+				if err := r.Release(data.clusterCIDR, cidr); err != nil {
+					return fmt.Errorf("failed to release cidr %q from clusterCIDR %q for node: %q: %w", cidr, data.clusterCIDR.Name, node.Name, err)
+				}
+			}
+		}
+		return err
+	}(data)
+
+	r.removeNodeFromProcessing(data.nodeName)
+	return err
+}
+
+// defaultNodeSelector generates a label for the catch all use case.
+func defaultNodeSelector() ([]byte, error) {
+	nodeSelector := &v1.NodeSelector{
+		NodeSelectorTerms: []v1.NodeSelectorTerm{
+			{
+				MatchExpressions: []v1.NodeSelectorRequirement{
+					{
+						Key:      defaultClusterCIDRKey,
+						Operator: v1.NodeSelectorOpIn,
+						Values:   []string{defaultClusterCIDRValue},
+					},
+				},
+			},
+		},
+	}
+
+	marshalledSelector, err := nodeSelector.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	return marshalledSelector, nil
+}
+
+// prioritizedCIDRs returns a list of CIDRs to be allocated to the node.
+// Returns 1 CIDR  if single stack.
+// Returns 2 CIDRs , 1 from each ip family if dual stack.
+func (r *multiCIDRRangeAllocator) prioritizedCIDRs(node *v1.Node) ([]*net.IPNet, *cidrset.ClusterCIDR, error) {
+	clusterCIDRList, err := r.orderedMatchingClusterCIDRs(node)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to get a clusterCIDR for node %s: %w", node.Name, err)
+	}
+
+	for _, clusterCIDR := range clusterCIDRList {
+		cidrs := make([]*net.IPNet, 0)
+		if clusterCIDR.IPv4CIDRSet != nil {
+			cidr, err := r.allocateCIDR(clusterCIDR, clusterCIDR.IPv4CIDRSet)
+			if err != nil {
+				klog.V(3).Infof("unable to allocate IPv4 CIDR, trying next range: %w", err)
+				continue
+			}
+			cidrs = append(cidrs, cidr)
+		}
+
+		if clusterCIDR.IPv6CIDRSet != nil {
+			cidr, err := r.allocateCIDR(clusterCIDR, clusterCIDR.IPv6CIDRSet)
+			if err != nil {
+				klog.V(3).Infof("unable to allocate IPv6 CIDR, trying next range: %w", err)
+				continue
+			}
+			cidrs = append(cidrs, cidr)
+		}
+
+		return cidrs, clusterCIDR, nil
+	}
+	return nil, nil, fmt.Errorf("unable to get a clusterCIDR for node %s, no available CIDRs", node.Name)
+}
+
+func (r *multiCIDRRangeAllocator) allocateCIDR(clusterCIDR *cidrset.ClusterCIDR, cidrSet *cidrset.MultiCIDRSet) (*net.IPNet, error) {
+	//r.lock.Lock()
+	//defer r.lock.Unlock()
+
+	for evaluated := 0; evaluated < cidrSet.MaxCIDRs; evaluated++ {
+		candidate, lastEvaluated, err := cidrSet.NextCandidate()
+		if err != nil {
+			return nil, err
+		}
+
+		evaluated += lastEvaluated
+
+		if r.cidrInAllocatedList(candidate) {
+			continue
+		}
+
+		// Deep Check.
+		if r.cidrOverlapWithAllocatedList(candidate) {
+			continue
+		}
+
+		// Mark the CIDR as occupied in the map.
+		if err := r.Occupy(clusterCIDR, candidate); err != nil {
+			return nil, err
+		}
```

Comment:
this operations is not locked, since NextCandidate to Occupy we can race, added several comments, these are just self notes for future reviews :)

Reply:
the caller of allocateCIDR acquires the lock in range allocator avoiding the race on the cidrMap where the cidrSets are stored: https://github.com/sarveshr7/kubernetes/blob/4316221e48ca90d28a01b867ad7adb76520c3b5c/pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go#L468